### ISSUE TITLE
[MPS] Severe Performance Degradation with Deterministic Algorithm on macOS 14.4

### ISSUE THREAD

**Issue by 1155159410** [pytorch/pytorch#122394]
## Issue Description
After updating to macOS 14.4, I have encountered significant performance degradation when using MPS with the deterministic algorithm **enabled**.

## Reproduce the Issue
Below is a minimal piece of code to reproduce the issue:
```python
import torch
import torchvision.models as models
from tqdm import tqdm

torch.manual_seed(0)
device = 'mps'

data = torch.rand(256, 3, 32, 32, device=device)  # Simulating a batch of 256, 32x32 RGB images
model = models.resnet152().to(device)  # Other models can also reproduce the issue (e.g., DenseNet, EfficientNet)

# Default, use_deterministic_algorithms == False
for i in tqdm(range(20)):
    model(data)

# use_deterministic_algorithms == True
torch.use_deterministic_algorithms(True)
for i in tqdm(range(20)):
    model(data)
```

## Current Behavior
Based on the result from my device (M1 Max 24-core GPU), the time used for the above forward pass with the **_deterministic algorithm disabled_** (default of PyTorch) and **_enabled_** are **_3s and 22s respectively_**. In other words, the degradation is roughly 8x. The result may be a bit different with different models, but still a severe degradation.

## Expected Behavior
It should be only around 2x degradation.

Since I identified this issue after upgrading to macOS 14.4, I didn't test the above code on macOS 14.3. However, I have some records of the performance on macOS 14.3 from a personal project. It shows that, for that project, training one epoch took around 85s without enabling the deterministic algorithm; and around 144s with deterministic algorithm enabled. Thus, the record suggested the degradation should be around 2x normally.

## Notes
From the "Activity Monitor", we can observe that **the GPU utilization caps at around 50% when the deterministic algorithm is enabled**. This cap persists regardless of the number of code instances executed concurrently, as long as they all have the deterministic algorithm enabled. (Total GPU utilization of all instances caps at around 50%) At the same time, the CPU utilization for each instance hits 100%. It appears that this behavior is the direct cause of this issue.

##
Thank you for your attention to this matter.

### Versions

PyTorch version: 2.2.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.4 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.2 | packaged by Anaconda, Inc. | (main, Feb 27 2024, 12:57:28) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-14.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.2.1
[pip3] torchaudio==2.2.1
[pip3] torchvision==0.17.1
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] torch                     2.2.1                    pypi_0    pypi
[conda] torchaudio                2.2.1                    pypi_0    pypi
[conda] torchvision               0.17.1                   pypi_0    pypi

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr

**Comment by malfet** [comment]
@1155159410 just to confirm, you don't see perf regression if deterministic algorithm is not use, do you?

**Comment by 1155159410** [comment]
@malfet Yes, there is no noticeable reduction in performance if the deterministic algorithm is not used (i.e., ```torch.use_deterministic_algorithms(False)```). Additionally, GPU utilization can be maxed out (reaching over 90%) when running the above reproduction code without the deterministic algorithm enabled. (Increasing the number of iterations may make it easier to observe.)

**Comment by malfet** [comment]
Ok, deterministic algorithms obviously would affect resize/empty as well as Indexing ops
https://github.com/pytorch/pytorch/blob/c2651a7f0e175e98b1f5f7ae65c993101fa86f4b/aten/src/ATen/native/mps/operations/Indexing.mm#L82
https://github.com/pytorch/pytorch/blob/c2651a7f0e175e98b1f5f7ae65c993101fa86f4b/aten/src/ATen/native/mps/TensorFactory.cpp#L123-L124

**Comment by kulinseth** [comment]
@1155159410 , did you run the same network on 14.3 vs 14.4 where the Deterministic / non-deterministic algorithms performance dropped ? I am curious because we have a single-threaded overly aggressive  kernel implemented for deterministic algorithm . I would have expected similar performance across 14.3 vs 14.4

**Comment by 1155159410** [comment]
@kulinseth Yes, I am sure that I was running the same script all the time before and after I upgraded to macOS 14.4.

Indeed, enabling the deterministic algorithm would affect the performance, but it was not this significant previously on macOS 14.3 (previously around 2x slower, currently around 8x slower).

Unfortunately, I am not able to downgrade to macOS 14.3 to perform further comparisons. I would appreciate it if someone with an older version of macOS could run the above reproduction code to see how much the performance reduction is given by the deterministic algorithm.

**Comment by 1155159410** [comment]
I found a friend willing to give a hand, with an Apple Silicon (M2) MacBook that runs an older version of macOS (13.1) to execute the reproduction code mentioned above. Here are the results:

- **Without the deterministic algorithm enabled** (PyTorch's default): Approximately 7 seconds (2.7 iter/s)
- **With the deterministic algorithm enabled**: Performance was slightly reduced, but still approximately 7 seconds (2.6 iter/s)

Although the M2's performance differs from that of my original setup (M1 Max), regarding "the performance when using deterministic algorithms", the results show only a small reduction. This suggests that the issue is probably indeed correlated with the macOS version. But it's unclear whether the root cause is an issue within macOS 14.4 itself or something that can be addressed on the PyTorch side.

Here are the details of the test environment:
```
PyTorch version: 2.2.2
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: Could not collect
Libc version: N/A

Python version: 3.12.2 (main, Feb  6 2024, 20:19:44) [Clang 15.0.0 (clang-1500.1.0.2.5)] (64-bit runtime)
Python platform: macOS-13.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.2.2
[pip3] torchaudio==2.2.2
[pip3] torchvision==0.17.2
[conda] numpy 1.26.4 pypi_0 pypi
[conda] numpydoc 1.5.0 py311hca03da5_0
```

**Comment by laclouis5** [comment]
I'm facing the same issue but I found that with PyTorch 2.0.1 the issue is not present. I got almost the same performance with and without deterministic algorithms enabled.

**Comment by kulinseth** [comment]
I will follow up to see if there are changes in Metal framework or other components in OS which are affecting this.

**Comment by 1155159410** [comment]
@laclouis5 Thank you for your information. I have tested on my device again and got same observations.

Besides, I also tried PyTorch 2.1.0 (the next version of 2.0.1), and it turned out that this issue can also be observed on 2.1.0. Here below are the results:

**PyTorch 2.0.1:**
* w/o deterministic: 2s, 7.06 it/s
* w/ deterministic: 2s, 7.30 it/s


**PyTorch 2.1.0:**
* w/o deterministic: 2s, 6.74it/s
* w/ deterministic: 23s, 1.20s/it

(All tests are conducted on macOS 14.4.1.)

**Comment by skotapati** [comment]
Unable to repro with latest nightly torch

**Comment by malfet** [comment]
> Unable to repro with latest nightly torch

Can you please post results here? I.e. in torch-2.0 numbers are X, in torch-2.4(or 5.rc1) numbers are Y?
